### PR TITLE
Upgrade es7 service

### PR DESCRIPTION
**Pull request by pkfec**
## Summary (required)

- Resolves https://github.com/fecgov/openFEC/issues/4505

**Action items**
- [x] Create a new ES7.4 High Availability beta service on `feature` space.
- [x] Set up API app on `feature` space.
- [x] Setup circleci env variable and deploy api on feature space successfully.
- [x] Refactor all mappings to match the new ES7 service (docs, archived_mur indexes).
- [x] Add new ES client connection function using AWS4Auth and test the ES connection.
- [x] Upload all legal docs to `feature` space.(AO, ADR, AF, MUR, Archived MURs, Statues, Regulations)
- [x] Refactor  existing index_management.py and move all the manage functions into the new file es_management.py.
- [x] Create new manage command (check_legal_doc) to view the legal data on the ES service.
- [x] Create a public S3 bucket service (fec-s3-api) to save legal pdf file. 
- [x] Configure the public S3 bucket to list the object in a browser (cloud.gov put the listobject policy  on this new bucket)
- [x] Create a new private s3 bucket service (fec-s3-snapshot) to save ES snapshot.(separated from public s3 bucket)
- [x] Config ES7 repository and  test snapshots 
- [x] Clean OCR text data on the **arch-mur253**(with 8 documents), which fixed the **Message: Request size exceeded 10485760 bytes** error while loading to ES7 service.
- [x] Refactor two endpoints:  **legal/search** and **legal/doc** 
- [x] Integrate CMS and API and test the legal search.
- [x] ~Refactor and~ test downloads.
- [x] ~Refactor and~ test snapshot and restore.
- [x] Create snapshots for each indexes (docs, archived_murs) and restore one at a time (revisit the code and test the current feature) 
- [x] Refactor all management commands.
- [x] Test schedule task code.
- [x] pytest test case for two endpoints:`/legal/search/` and `/legal/docs/<doc_type>/<no>`
- [x] pytest test case for upload legal documents and es-management (in separated ticket)
[https://github.com/fecgov/openFEC/issues/4692](https://github.com/fecgov/openFEC/issues/4692)
- [x] Update ReadMe file (in separated ticket)
[https://github.com/fecgov/openFEC/issues/4693](https://github.com/fecgov/openFEC/issues/4693)
- [x] Conduct performance test (load test).https://docs.google.com/spreadsheets/d/1WEj_LeUgUyK0_9Kv_pWEIXCt76KMNZxQdfR8BMdk7W4/edit#gid=0

## How to test the changes locally
- [x] Part 1: test locally
1)checkout branch
2)Wiki instruction: [https://github.com/fecgov/openFEC/wiki/Elasticsearch-7.x.0-setup-and-test-locally](https://github.com/fecgov/openFEC/wiki/Elasticsearch-7.x.0-setup-and-test-locally)
- [x] Part 2: test on **feature** space
Wiki instruction: [https://github.com/fecgov/openFEC/wiki/Elasticsearch-7.x.0-management-instruction](https://github.com/fecgov/openFEC/wiki/Elasticsearch-7.x.0-management-instruction)
- [ ] Part 3: test on **dev** or **stage**
Wiki instruction: [https://github.com/fecgov/openFEC/wiki/Elasticsearch-7.x.0-service-and-private-s3-bucket-setup-on-'space'](https://github.com/fecgov/openFEC/wiki/Elasticsearch-7.x.0-service-and-private-s3-bucket-setup-on-'space')
1)Checkout this branch and create a new test branch 
2)private s3 bucket has already exist, try to connect to it (need this to view snapshots)
3)create a new ES7 service instance (15 mins)
4)create ES service key and bind private S3 bucket to ES7 service
5)get snapshotRoleARN and update env variable : **ES_SNAPSHOT_ROLE_ARN**
6)Rename the existing public s3 bucket service name from **fec-s3-<space>** to **fec-s3-api**, 
   To rename run this command - `cf rename-service fec-s3-<space> fec-s3-api`
7)Set Pingdom maintenance window
8)unbind api, celery-worker, celery-beat from ES56 service - 
`cf unbind-service api fec-api-search56`
`cf unbind-service celery-beat fec-api-search56`
`cf unbind-service celery-worker fec-api-search56`
9)deploy your test branch to circleci or `invoke deploy --space <space>
10)upload the legal data to 'docs' index - 
`cf run-task api --command "python manage.py refresh_current_legal_docs_zero_downtime" -m 4G --name refresh_data`
11)upload the archived murs data
`cf run-task api --command "python manage.py initialize_archived_mur_docs" -m 4G --name initialize_arch_mur_data`
12)create repository and take a new snapshot and restore
13)test legal endpoint
[https://fec-dev-api.app.cloud.gov/v1/legal/search/](https://fec-dev-api.app.cloud.gov/v1/legal/search/)
[https://api-stage.open.fec.gov/v1/legal/search/?api_key=DEMO_KEY](https://api-stage.open.fec.gov/v1/legal/search/?api_key=DEMO_KEY)
[https://api.open.fec.gov/v1/legal/search/?api_key=DEMO_KEY](https://api.open.fec.gov/v1/legal/search/?api_key=DEMO_KEY)
14)test cms legal search page

- [ ] Part 4: put back ES56 (optional)
1)rename public s3 service
`cf rename-service fec-s3-api fec-s3-<space>`
2)rebuild the latest api on circleci
3)unbind ES74 (manually)
`cf unbind-service api fec-api-elasticsearch`
`cf unbind-service celery-beat fec-api-elasticsearch`
`cf unbind-service celery-worker fec-api-elasticsearch`
4)delete ES7 service
`cf delete-service-key fec-api-elasticsearch fec-api-elasticsearch-key`
`cf delete-service fec-api-elasticsearch`
5)check legal endpoint and cms page

## Reviewer
Laura is mandatory and Welcome everyone.

## Impacted areas of the application
List general components of the application that this PR will affect:

1. Legal Resource on CMS
2. Legal API endpoints

## Related PRs
List related PRs https://github.com/fecgov/openFEC/pull/4596

### CODE REVIEW

**Comment by fec-jli** [comment]
====Refactor **legal search** and **legal doc** two endpoints====
ES7 breaking changes
https://www.elastic.co/guide/en/elasticsearch/reference/current/breaking-changes-7.0.html

1)remove mapping type ( _type)  under index level, default always is _doc now
replace by adding field in property: type
[https://www.elastic.co/guide/en/elasticsearch/reference/current/removal-of-types.html](https://www.elastic.co/guide/en/elasticsearch/reference/current/removal-of-types.html)
Screen Shot 2020-10-07 at 5 09 51 PM

**Comment by pkfec** [comment]
Refactored `test_load_legal_docs.py`. 7 tests pass only when the ES74 service is UP and RUNNING on local env, unlike ES56. Observed that tests are failing on circleci for the same reason. 

**Error stack trace:**
   _urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x10b4d0690>: Failed to establish a new connection: [Errno 61] Connection refused_

File: requirements.txt [hunk 1]
```diff
@@ -23,8 +23,11 @@ greenlet==0.4.16 # pinned to fix build problem (parent is gevent)
 webargs==5.5.3
 ujson==1.33
 requests==2.22.0
-elasticsearch==5.5.3
-elasticsearch-dsl==5.4.0
+# elasticsearch==5.5.3
+# elasticsearch-dsl==5.4.0
```
Comment (lbeaufort) [inline requirements.txt:27]: Can you remove the old dependencies please?

File: webservices/legal_docs/archived_murs.py [hunk 2]
```diff
@@ -3,18 +3,22 @@
...
+    ARCHIVED_MURS_INDEX,
+    ARCHIVED_MURS_ALIAS,
+)
 
 logger = logging.getLogger(__name__)
 
-# for debug
-# logger.setLevel(logging.DEBUG)
+# for debug, uncomment this line
+logger.setLevel(logging.DEBUG)
```
Comment (lbeaufort) [inline webservices/legal_docs/archived_murs.py:21]: Should this be commented before merge?

File: webservices/legal_docs/current_cases.py [hunk 3]
```diff
@@ -362,10 +401,10 @@ def get_sorted_respondents(participants):
     """
     Returns the respondents in a MUR sorted in the order of most important to least important
     """
-    SORTED_RESPONDENT_ROLES = ['Primary Respondent', 'Respondent', 'Previous Respondent']
+    SORTED_RESPONDENT_ROLES = ["Primary Respondent", "Responden", "Previous Respondent"]
```
Comment (lbeaufort) [inline webservices/legal_docs/current_cases.py:404]: Is this missing a `t`?

` SORTED_RESPONDENT_ROLES = ["Primary Respondent", "Responden", "Previous Respondent"]`

File: webservices/tasks/legal_docs.py [hunk 4]
```diff
@@ -38,6 +38,9 @@
     ORDER BY case_serial
 """
 
+# SLACK_BOTS = "#bots"
+SLACK_BOTS = "#test-bot"
```
Comment (lbeaufort) [inline webservices/tasks/legal_docs.py:42]: Should this be `"#bots"` before merge?

File: webservices/tasks/__init__.py [hunk 5]
```diff
@@ -8,34 +8,34 @@
...
-            'schedule': crontab(minute='*/5', hour='10-23'),
-        },
-
-        'backup_elasticsearch_every_sunday': {
-            'task': 'webservices.tasks.legal_docs.create_es_backup',
-            'schedule': crontab(minute=0, hour=4, day_of_week='sun'),
-        },
+# if env.app.get('space_name', 'unknown-space').lower() != 'feature':
+schedule = {
+    # 'refresh_materialized_views': {
```
Comment (lbeaufort) [inline webservices/tasks/__init__.py:13]: I think this needs to be uncommented before merge.

File: webservices/utils.py [hunk 6]
```diff
@@ -442,8 +448,59 @@ def get_current_cycle():
...
+                retry_on_timeout=True,
+            )
+        return es_client
+    except Exception as err:
+        logger.error('An error occurred trying to create Elasticsearch client.{0}'.format(err))
+
+
 def get_elasticsearch_connection():
-    es_conn = env.get_service(name='fec-api-search56')
+    es_conn = env.get_service(name='fec-api-elasticsearch')
```
Comment (lbeaufort) [inline webservices/utils.py:503]: Should this use `ES_SERVICE_INSTANCE_NAME`?

File: webservices/tasks/legal_docs.py (hunk 4 above)
Comment (pkfec) [reply webservices/tasks/legal_docs.py:42]: Yes, will revert to #bots once testing is done.

File: webservices/tasks/__init__.py (hunk 5 above)
Comment (pkfec) [reply webservices/tasks/__init__.py:13]: thats correct, will revert this change!

File: webservices/legal_docs/current_cases.py (hunk 3 above)
Comment (pkfec) [reply webservices/legal_docs/current_cases.py:404]: good catch. Will fix the spelling

File: webservices/utils.py (hunk 6 above)
Comment (pkfec) [reply webservices/utils.py:503]: This is obsolete code forgot to cleanup. Thanks for flagging.

File: webservices/legal_docs/archived_murs.py (hunk 2 above)
Comment (pkfec) [reply webservices/legal_docs/archived_murs.py:21]: Yes, will make the change.

File: requirements.txt (hunk 1 above)
Comment (pkfec) [reply requirements.txt:27]: Obsolete code, forgot to cleanup. Thanks for flagging.

**Comment by fec-jli** [comment]
@lbeaufort Thank you review this. We have updated all the changes you point out and pushed.

File: webservices/resources/legal.py [hunk 7]
```diff
@@ -121,27 +98,30 @@ def get(self, q='', from_hit=0, hits_returned=20, **kwargs):
...
-            doc_types = [kwargs.get('type')]
+            doc_types = [kwargs.get("type")]
 
             # if doc_types is not in one of ALL_DOCUMENT_TYPES
             # then reset type = all (= ALL_DOCUMENT_TYPES)
             if doc_types[0] not in ALL_DOCUMENT_TYPES:
                 doc_types = ALL_DOCUMENT_TYPES
 
         hits_returned = min([200, hits_returned])
+        print(hits_returned)
```
Comment (lbeaufort) [inline webservices/resources/legal.py:112]: I think this `print` statement needs to be removed

File: webservices/utils.py [hunk 8]
```diff
@@ -13,7 +13,9 @@
 from sqlalchemy.orm import foreign
 from sqlalchemy.ext.declarative import declared_attr
 from sqlalchemy.dialects import postgresql
-from elasticsearch import Elasticsearch
+from elasticsearch import Elasticsearch, RequestsHttpConnection
+import certifi
```
Comment (lbeaufort) [inline webservices/utils.py:17]: Does `certifi` need to be in `requirements.txt`?

**Review by lbeaufort — APPROVED** [review]
Excellent work and documentation, @pkfec and @fec-jli! Just two small comments ⬆️

File: webservices/resources/legal.py (hunk 7 above)
Comment (pkfec) [reply webservices/resources/legal.py:112]: Thanks for flagging. Will remove this line.

File: webservices/utils.py (hunk 8 above)
Comment (pkfec) [reply webservices/utils.py:17]: certifi is a dependency pkg of requests-aws4auth==1.0 which we already have in requirements.txt file. We don't need to to specify again.

Comment (lbeaufort) [reply webservices/utils.py:17]: Because we're using it as a standalone package, let's go ahead and put it in requirements.txt to avoid breaking changes.

Comment (pkfec) [reply webservices/utils.py:17]: done!

Comment (lbeaufort) [reply webservices/utils.py:17]: Thanks @pkfec! Interesting versioning `certifi` has - looks like they're just tagging releases with dates and not following semantic versioning...

Thanks for addressing my comments! Looks great.

**Review by lbeaufort — APPROVED** [review]
Looks great! Thanks for making all the changes 👍

**Pull request merged**